### ISSUE TITLE
Add default editor templates for all numeric types, including short, ushort, and double, etc.

### ISSUE THREAD

**Issue by weitzhandler** [aspnet/Mvc#2261]
Hi.

Please have a look at [this](https://aspnetwebstack.codeplex.com/workitem/2) post.

It addresses what the `System.ComponentModel.DataAnnotations.DataType` attribute is here for.

Anyway, as per the [closing note](https://aspnetwebstack.codeplex.com/workitem/2#ClosedDiv), only the following numeric types are supported:
>type=number
int
byte
sbyte
uint
long
ulong
DataType enum
PhoneNumber (tel)
EmailAddress (email)
Url (url)
Date (date)
Time (time)
DateTime (datetime)

I'd like to ask you to please include short in the numeric category as well.

**Comment by danroth27** [comment]
@dougbu Can you check if short is actually supported?

**Comment by dougbu** [comment]
@weitzhandler server-side we simply execute any `ValidationAttribute` or subclass we find.  client side we do two things:
1. special-case some attributes for Unobtrusive client-side validation e.g. `[EmailAddress]` and `[Compare]`
2. if you use `@Html.Editor()`, `@Html.EditorFor()`, or the `<input/>` tag helper, generate `<input/>` elements with the appropriate `type` attribute based on `[DataType]`.  this allows browsers to validate the inputs.

I suspect the scenario you've opened this issue about is 2. in my list -- where we generate `<input type="number" .../>` and where we don't.  your `type=number` list above matches what we have in both MVC 5 and 6.

please confirm your request is for helper calls such as `@Html.EditorFor(m => m.ShortProperty)` to generate `<input type="number" .../>`.

**Comment by weitzhandler** [comment]
@dougbu 
Yes that's exactly what I meant.

**Comment by dougbu** [comment]
updated subject to reflect what would need to change here.

@weitzhandler In the meantime you could create a Shared/EditorTemplates/Short.cshtml file containing for example
``` c#
@model short
@Html.TextBoxFor(m => m, htmlAttributes: new { type = "number" })
```

**Comment by weitzhandler** [comment]
I'd also recommend adding those templates for `ushort` and all the other CLR primitive numeric types.

**Comment by weitzhandler** [comment]
@Eilon and @dougbu  [Here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs)'s your friend. And BTW, I don't see any reason not to have `decimal` as an `<input type="number" />` as well.

As a temporary workaround, I'd like to add *Int16.cshtml* to *~/Views/Shared/EditorTemplates*, but I'm not sure what I should include in the template, since I want the validations to be generated as well.

**Comment by dougbu** [comment]
> not sure what I should include in the template, since I want the validations to be generated as well

do you mean validation attributes?  should get those automatically with what I recommended:
``` c#
@model short
@Html.TextBoxFor(m => m, htmlAttributes: new { type = "number" })
```

or, if you prefer tag helpers
``` c#
@model short
@addTagHelper "*, Microsoft.AspNet.Mvc.TagHelpers"
<input asp-for="@Model" type="number"/>
```

otherwise feel free to include `@Html.ValidationMessageFor(...)`.  one downside is that would be very inconsistent with the default templates.

> don't see any reason not to have `decimal`

`decimal` is already special-cased w.r.t. formatting.  but you are correct that the `<input/>` tag helper and `@Html.Editor()` currently use the basic `type="text"` for that type.

**Comment by weitzhandler** [comment]
As you can see @dougbu, I've already deleted the 2nd half of your question, I missed what you wrote about using the `TextBoxFor` in [this post](https://github.com/aspnet/Mvc/issues/2261#issuecomment-86632052).

Thanks!

**Comment by weitzhandler** [comment]
As you can see @dougbu, I've already deleted the 2nd half of your question, I missed what you wrote about using the `TextBoxFor` in [this post](https://github.com/aspnet/Mvc/issues/2261#issuecomment-86632052).

I'm sure you're responding by email, so you didn't get the updated version...

Thanks!

**Comment by dougbu** [comment]
fixed in 73e8fc1
